compiler/circle-mpqsolver/src/bisection/BisectionSolver.cpp

```
@@ -0,0 +1,227 @@
+{
+  auto model = read_module(flt_path);
+  // get fake quantized model for evaluation
+  if (!_quantizer->fake_quantize(model.get(), def_quant, layers))
+  {
+    throw std::runtime_error("Failed to produce fake-quantized model.");
+  }
+
+  return evaluator.evaluate(model.get());
+}
+
+std::unique_ptr<luci::Module> BisectionSolver::run(const std::string &module_path)
+{
+  LOGGER(l);
+
+  auto module = read_module(module_path);
+
+  float min_depth = 0.f;
+  float max_depth = 0.f;
+  NodeDepthType nodes_depth;
+  if (compute_depth(module.get(), nodes_depth, min_depth, max_depth) !=
+      ParameterizerResult::SUCCESS)
+  {
+    std::cerr << "Invalid graph for bisectioning" << std::endl;
+    return nullptr;
+  }
+
+  std::unique_ptr<MAEMetric> metric = std::make_unique<MAEMetric>();
+  DatasetEvaluator evaluator(module.get(), _input_data_path, *metric.get());
+
+  // const auto &ref_output = compute_outputs(module.get(), _input_data_path);
+  LayerParams layer_params;
+  float int16_qerror = evaluate(evaluator, module_path, "int16", layer_params);
+  VERBOSE(l, 0) << "int16_quantization_error " << int16_qerror << std::endl;
+
+  float int8_qerror = evaluate(evaluator, module_path, "uint8", layer_params);
+  VERBOSE(l, 0) << "int8_quantization_error " << int8_qerror << std::endl;
+
+  _qerror = int16_qerror + _qerror_ratio * std::fabs(int8_qerror - int16_qerror);
+  VERBOSE(l, 0) << "target quantization error " << _qerror << std::endl;
+
+  if (int16_qerror > _qerror)
+  {
+    std::cerr << "ERROR: minimal targeted is an error of Q16 quantized model" << std::endl;
+    return nullptr;
+  }
+  if (int8_qerror < _qerror)
+  {
+    // no need for bisectioning just return Q8 model
+    if (!_quantizer->quantize(module.get(), "uint8", layer_params))
+    {
+      std::cerr << "ERROR: failed to quantize model" << std::endl;
+      return nullptr;
+    }
+  }
+
+  int last_depth = -1;
+  float best_depth = -1;
+  LayerParams best_params;
+  auto graph = module->graph(0);
```

Please throw an exception if module has multiple subgraphs.

---

ok